src/PostgresOutbox/Subscriptions/Subscription.cs

```diff
@@ -0,0 +1,160 @@
+            await foreach (var subscribe in ProcessEnvelope<IEnvelope>(envelope, registry, errorProcessor).WithCancellation(ct))
+                yield return subscribe;
+        }
+
+        await foreach (var message in
+                       _connection.StartReplication(slot,
+                           new PgOutputReplicationOptions(publicationSetupOptions.PublicationName, 1, slotSetupOptions.Binary), ct))
+        {
+            if (message is InsertMessage insertMessage)
+            {
+                var envelope = await replicationDataMapper.ReadFromReplication(insertMessage, ct);
+                await foreach (var subscribe in ProcessEnvelope<IEnvelope>(envelope, registry, errorProcessor).WithCancellation(ct))
+                    yield return subscribe;
+            }
+            // Always call SetReplicationStatus() or assign LastAppliedLsn and LastFlushedLsn individually
+            // so that Npgsql can inform the server which WAL files can be removed/recycled.
+            _connection.SetReplicationStatus(message.WalEnd);
+            await _connection.SendStatusUpdate(ct);
+        }
+    }
+
+    private static async IAsyncEnumerable<T> ProcessEnvelope<T>(
+        IEnvelope envelope,
+        Dictionary<Type, IConsume> registry,
+        IErrorProcessor errorProcessor
+    ) where T:class
+    {
+        switch (envelope)
+        {
+            case KoEnvelope error:
+                await errorProcessor.Process(error.Error);
+                yield break;
+            case OkEnvelope okEnvelope:
+            {
+                var obj = okEnvelope.Value;
+                var objType = obj.GetType();
+                var (consumer, methodInfo) = Memoize(registry, objType, Consumer);
+                await ((Task)methodInfo.Invoke(consumer, [obj])!).ConfigureAwait(false);
+                yield return (T)envelope;
+                yield break;
+            }
+        }
+    }
+
+    private static readonly Dictionary<Type, (IConsume consumer, MethodInfo methodInfo)> Cache = [];
+    
+
+    private static (IConsume consumer, MethodInfo methodInfo) Memoize
+    (
+        Dictionary<Type, IConsume> registry,
+        Type objType,
+        Func<Dictionary<Type, IConsume>, Type, (IConsume consumer, MethodInfo methodInfo)> func
+    )
+    {
+        if (!Cache.TryGetValue(objType, out var entry))
+            entry = func(registry, objType);
+        Cache[objType] = entry;
+        return entry;
+    }
+    private static (IConsume consumer, MethodInfo methodInfo) Consumer(Dictionary<Type, IConsume> registry, Type objType)
+    {
+        var consumer = registry[objType] ?? throw new NotSupportedException($"Unregistered type for {objType.AssemblyQualifiedName}");
+        var methodInfos = consumer.GetType().GetMethods(BindingFlags.Instance|BindingFlags.Public);
+        var methodInfo = methodInfos?.SingleOrDefault(mi=>mi.GetParameters().Any(pa => pa.ParameterType == objType))
```

Comment:
**CONCERN:** Is this used to call the generic method of the `IConsume`? If yes, then I'd recommend extending IConsume to have a method accepting an object message, then default implementation in IConsume<T> that would call the typed method. This would reduce the need for this reflection access here, we could just call it.

Reply:
We might support different scenario.
A truly confident user could simply decide to deal with raw data, hence a basic `IConsume` would be enough. Or he would prefer to work typed data and generics support, hence some sort of (memoized)reflections stuff might be unavoidable. Some benchmark here might help in solving the dilemma of reflection adoption.

Reply:
Added: https://github.com/oskardudycz/PostgresOutboxPatternWithCDC.NET/issues/17